net/httpclient.go

```diff
@@ -2,18 +2,219 @@ package net
 
 import (
 	"crypto/tls"
+	"io"
 	"net/http"
 	"net/http/httptrace"
+	"net/url"
+	"strings"
 	"time"
 
 	"github.com/opentracing/opentracing-go"
 	"github.com/opentracing/opentracing-go/ext"
+	"github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/logging"
+	"github.com/zalando/skipper/secrets"
 )
 
 const (
 	defaultIdleConnTimeout = 30 * time.Second
+	defaultRefreshInterval = 5 * time.Minute
 )
 
+// StaticSecret implements SecretsReader interface. Example:
+//
+//    sec := []byte("mysecret")
+//    sss := StaticSecret(sec)
+//    b,_ := sss.GetSecret("")
+//    string(b) == sec // true
+type StaticSecret []byte
+
+// GetSecret returns the static secret
+func (st StaticSecret) GetSecret(string) ([]byte, bool) {
+	return st, true
+}
+
+// Close implements secrets.SecretsReader.
+func (st StaticSecret) Close() {}
+
+// StaticDelegateSecret delegates with a static string to the wrapped
+// SecretsReader
+type StaticDelegateSecret struct {
+	sr  secrets.SecretsReader
+	key string
+}
+
+// NewStaticDelegateSecret creates a wrapped secrets.SecretsReader,
+// that use given s to the underlying secrets.SecretsReader to return
+// the secret.
+func NewStaticDelegateSecret(sr secrets.SecretsReader, s string) *StaticDelegateSecret {
+	return &StaticDelegateSecret{
+		sr:  sr,
+		key: s,
+	}
+}
+
+// GetSecret returns the secret looked up by the static key via
+// delegated secrets.SecretsReader.
+func (sds *StaticDelegateSecret) GetSecret(string) ([]byte, bool) {
+	return sds.sr.GetSecret(sds.key)
+}
+
+// Close delegates to the wrapped secrets.SecretsReader.
+func (sds *StaticDelegateSecret) Close() {
+	sds.sr.Close()
+}
+
+// HostSecret can be used to get secrets by hostnames.
+type HostSecret struct {
+	sr     secrets.SecretsReader
+	secMap map[string]string
+}
+
+// NewHostSecret create a SecretsReader that returns a secret for
+// given host. The given map is used to map hostname to the secrets
+// reader key to read the secret from.
+func NewHostSecret(sr secrets.SecretsReader, h map[string]string) *HostSecret {
+	if h == nil {
+		h = make(map[string]string)
+	}
+	return &HostSecret{
+		sr:     sr,
+		secMap: h,
+	}
+}
+
+// GetSecret returns secret for given URL string using the hostname.
+func (hs *HostSecret) GetSecret(s string) ([]byte, bool) {
+	u, err := url.Parse(s)
+	if err != nil {
+		return nil, false
+	}
+	hostname := u.Hostname()
+	k, ok := hs.secMap[hostname]
+	if !ok {
+		return nil, false
+	}
+	b, ok := hs.sr.GetSecret(k)
+	if !ok {
+		return nil, false
+	}
+	return b, true
+}
+
+func (hs *HostSecret) Close() {
+	hs.sr.Close()
+}
+
+// Client adds additional features like Bearer token injection, and
+// opentracing to the wrapped http.Client with the same interface as
+// http.Client from the stdlib.
+type Client struct {
+	client http.Client
+	tr     *Transport
+	log    logging.Logger
+	sr     secrets.SecretsReader
+	quit   chan struct{}
```

Comment:
i think we don't need this quit attribute

Reply:
Thanks! Cleaned up this was used by an experiment